src/stages/main/patchers/WhilePatcher.js

```diff
@@ -125,9 +126,16 @@ export default class WhilePatcher extends NodePatcher {
   patchAsExpression() {
     this.body.setImplicitlyReturns();
     let resultBinding = this.getResultArrayBinding();
-    this.insert(this.contentStart, `(() => { ${resultBinding} = []; `);
     this.patchAsStatement();
-    this.insert(this.contentEnd, ` return ${resultBinding}; })()`);
+    let prefix = this.yielding ? 'yield* (function*()' : '(() =>';
+    this.insert(this.contentStart, `${prefix} { ${resultBinding} = []; `);
+    let postfix = this.yielding ? '.call(this)' : '()';
```

Comment:
We should change `.call(this)` to `.apply(this, arguments)`, since the `while` body might reference `arguments`, which would refer to a different value inside a generator function.